File: apps/android/app/src/main/java/com/galaxyssi/chat/blob/BlobArtifactContract.kt
package com.galaxyssi.chat.blob

import org.json.JSONObject
import java.net.URI
import java.util.Locale

/** Small authenticated output offers; file bytes never belong in this contract. */
internal object BlobArtifactContract {
    const val OFFER_TYPE = "artifact_blob_offer"
    const val RECEIPT_TYPE = "artifact_blob_receipt"
    const val MAX_CONTROL_BYTES = 32 * 1024
    private val scope = listOf("client_route_id", "conversation_id", "task_id", "turn_id", "contact_id",
        "source_message_id", "desktop_id")
    private val textLimits = scope.associateWith { 256 } + mapOf("artifact_id" to 64, "artifact_uri" to 2048,
        "name" to 255, "relative_path" to 2048, "mime_type" to 255, "sha256" to 64, "original_sha256" to 64)
    private val numberLimits = mapOf("size_bytes" to BlobProtocol.MAX_FILE_BYTES,
        "original_size_bytes" to 9_007_199_254_740_991L, "execution_generation" to 9_007_199_254_740_991L)
    private val fields = textLimits.keys + numberLimits.keys + "peer_chat"

    fun makeManifest(metadata: JSONObject): JSONObject {
        BlobProtocol.keys(metadata, fields)
        textLimits.forEach { (key, maximum) ->
            val text = BlobProtocol.string(metadata, key)
            if (text.codePointCount(0, text.length) !in 1..maximum || text.any { it.code < 32 }) {
                BlobProtocol.fail("invalid_artifact_blob_manifest")
            }
        }
        numberLimits.forEach { (key, maximum) -> BlobProtocol.integer(metadata, key, 1, maximum) }
        if (metadata.opt("peer_chat") !is Boolean ||
            !BlobProtocol.string(metadata, "client_route_id").matches(Regex("[A-Za-z0-9_-]{22}"))) {
            BlobProtocol.fail("invalid_artifact_blob_manifest")
        }
        listOf("artifact_id", "sha256", "original_sha256").forEach {
            BlobProtocol.unhex(BlobProtocol.string(metadata, it), 32).fill(0)
        }
        val uri = try { URI(BlobProtocol.string(metadata, "artifact_uri")) }
            catch (_: Exception) { BlobProtocol.fail("invalid_artifact_blob_uri") }
        if (uri.scheme?.lowercase(Locale.ROOT) != "galaxyssi-artifact" || uri.rawAuthority.isNullOrBlank() ||
            uri.rawQuery != null || uri.rawFragment != null) BlobProtocol.fail("invalid_artifact_blob_uri")
        if (metadata.getLong("original_size_bytes") < metadata.getLong("size_bytes")) {
            BlobProtocol.fail("invalid_artifact_blob_manifest")
        }
        val canonical = BlobProtocol.canonical(metadata)
        if (canonical.size > MAX_CONTROL_BYTES / 2) BlobProtocol.fail("artifact_blob_manifest_too_large")
        return JSONObject(metadata.toString()).put("transfer_id", BlobProtocol.hash(canonical))
    }

    fun validateManifest(manifest: JSONObject): JSONObject {
        BlobProtocol.keys(manifest, fields + "transfer_id")
        val transferId = BlobProtocol.string(manifest, "transfer_id").also { BlobProtocol.unhex(it, 32).fill(0) }
        val metadata = JSONObject(manifest.toString()).also { it.remove("transfer_id") }
        return makeManifest(metadata).also {
            if (it.getString("transfer_id") != transferId) BlobProtocol.fail("artifact_blob_manifest_mismatch")
        }
    }

    fun binding(manifest: JSONObject): Map<String, String> {
        val value = validateManifest(manifest)
        return (scope.associateWith { value.getString(it) } + mapOf("artifact_id" to value.getString("artifact_id"),
            "transfer_id" to value.getString("transfer_id"),
            "execution_generation" to value.getLong("execution_generation").toString()))
            .also { BlobProtocol.bindingHash(it) }
    }

    fun validateOffer(payload: JSONObject, route: String, desktop: String, origin: String): JSONObject {
        if (payload.optString("type") != OFFER_TYPE) BlobProtocol.fail("invalid_artifact_blob_offer")
        BlobProtocol.integer(payload, "version", 1, 1)
        val revision = BlobProtocol.integer(payload, "transport_revision", 1, 9_007_199_254_740_991L)
        if (BlobProtocol.canonical(payload).size > MAX_CONTROL_BYTES) BlobProtocol.fail("artifact_blob_offer_too_large")
        val manifest = validateManifest(payload.optJSONObject("manifest") ?: BlobProtocol.fail("invalid_artifact_blob_manifest"))
        if (manifest.getString("client_route_id") != route || manifest.getString("desktop_id") != desktop) {
            BlobProtocol.fail("artifact_blob_route_mismatch")
        }
        val offer = payload.optJSONObject("blob_offer") ?: BlobProtocol.fail("invalid_blob_offer")
        BlobProtocol.keys(offer, setOf("version", "relay", "private", "read_token"))
        BlobProtocol.integer(offer, "version", 1, 1)
        if (BlobProtocol.string(offer, "relay") != BlobHttp.normalizeOrigin(origin)) {
            BlobProtocol.fail("artifact_blob_relay_mismatch")
        }
        BlobProtocol.unhex(BlobProtocol.string(offer, "read_token"), 32).fill(0)
        val private = offer.optJSONObject("private") ?: BlobProtocol.fail("invalid_private_descriptor")
        BlobProtocol.keys(private, setOf("version", "blob_id", "key", "nonce_prefix", "size", "sha256",
            "binding_sha256", "manifest_sha256"))
        BlobProtocol.integer(private, "version", 1, 1)
        BlobProtocol.integer(private, "size", 0, BlobProtocol.MAX_FILE_BYTES)
        mapOf("blob_id" to 16, "key" to 32, "nonce_prefix" to 8, "sha256" to 32,
            "binding_sha256" to 32, "manifest_sha256" to 32).forEach { (key, bytes) ->
            BlobProtocol.unhex(BlobProtocol.string(private, key), bytes).fill(0)
        }
        if (private.getString("binding_sha256") != BlobProtocol.bindingHash(binding(manifest)) ||
            private.getLong("size") != manifest.getLong("size_bytes") ||
            private.getString("sha256") != manifest.getString("sha256")) {
            BlobProtocol.fail("artifact_blob_binding_mismatch")
        }
        return JSONObject().put("type", OFFER_TYPE).put("version", 1).put("manifest", manifest).put("transport_revision", revision)
            .put("blob_offer", JSONObject(offer.toString()))
    }

    fun storedReceipt(manifest: JSONObject): JSONObject {
        val value = validateManifest(manifest)
        return JSONObject().put("type", RECEIPT_TYPE).put("version", 1).put("status", "stored").also { receipt ->
            listOf("transfer_id", "client_route_id", "artifact_id", "sha256", "size_bytes").forEach {
                receipt.put(it, value.get(it))
            }
        }
    }
}

File: apps/desktop/core/galaxyssi-link/backend/blob_artifact_contract.py
"""Authenticated output-artifact metadata, independent of the bulk byte stream."""
from __future__ import annotations

from urllib.parse import urlsplit

from blob_client import relay_origin
from blob_crypto import binding_hash, validate_private_descriptor
from blob_protocol import BlobError, MAX_FILE_BYTES, canonical, checked_hex, sha256
from link_protocol import valid_route_id

OFFER_TYPE = "artifact_blob_offer"
RECEIPT_TYPE = "artifact_blob_receipt"
MAX_CONTROL_BYTES = 32 * 1024
SCOPE_FIELDS = ("client_route_id", "conversation_id", "task_id", "turn_id", "contact_id",
                "source_message_id", "desktop_id")
_TEXT_LIMITS = {**dict.fromkeys(SCOPE_FIELDS, 256), "artifact_id": 64, "artifact_uri": 2048,
                "name": 255, "relative_path": 2048, "mime_type": 255,
                "sha256": 64, "original_sha256": 64}
_NUMBER_LIMITS = {"size_bytes": MAX_FILE_BYTES, "original_size_bytes": 2**53 - 1,
                  "execution_generation": 2**53 - 1}
_FIELDS = set(_TEXT_LIMITS) | set(_NUMBER_LIMITS) | {"peer_chat"}


def _metadata(value: dict) -> dict:
    if not isinstance(value, dict) or set(value) != _FIELDS:
        raise BlobError("invalid_artifact_blob_manifest")
    for key, maximum in _TEXT_LIMITS.items():
        text = value[key]
        if (not isinstance(text, str) or not 1 <= len(text) <= maximum
                or any(ord(char) < 32 for char in text)):
            raise BlobError("invalid_artifact_blob_manifest")
    for key, maximum in _NUMBER_LIMITS.items():
        if type(value[key]) is not int or not 1 <= value[key] <= maximum:
            raise BlobError("invalid_artifact_blob_manifest")
    if type(value["peer_chat"]) is not bool or not valid_route_id(value["client_route_id"]):
        raise BlobError("invalid_artifact_blob_manifest")
    for key in ("artifact_id", "sha256", "original_sha256"):
        checked_hex(value[key])
    try:
        uri = urlsplit(value["artifact_uri"])
        if uri.scheme != "galaxyssi-artifact" or not uri.netloc or uri.query or uri.fragment:
            raise ValueError()
    except ValueError:
        raise BlobError("invalid_artifact_blob_uri") from None
    if value["original_size_bytes"] < value["size_bytes"]:
        raise BlobError("invalid_artifact_blob_manifest")
    if len(canonical(value)) > MAX_CONTROL_BYTES // 2:
        raise BlobError("artifact_blob_manifest_too_large", 413)
    return dict(value)


def make_manifest(metadata: dict) -> dict:
    value = _metadata(metadata)
    return {**value, "transfer_id": sha256(canonical(value))}


def make_scoped_manifest(metadata: dict) -> dict:
    """Build from the original prepared artifact, before publishing rich reply links.

    The same task path in another turn, device, or execution generation must not
    replace a previously displayed phone artifact. Retries of unchanged metadata
    deliberately keep the same URI and transfer identity.
    """
    value = _metadata(metadata)
    value["artifact_uri"] = "galaxyssi-artifact://blob/" + sha256(canonical(value))
    return make_manifest(value)


def validate_manifest(value: dict) -> dict:
    if not isinstance(value, dict) or set(value) != _FIELDS | {"transfer_id"}:
        raise BlobError("invalid_artifact_blob_manifest")
    expected = make_manifest({key: value[key] for key in _FIELDS})
    if checked_hex(value["transfer_id"]) != expected["transfer_id"]:
        raise BlobError("artifact_blob_manifest_mismatch", 409)
    return expected


def artifact_binding(manifest: dict) -> dict:
    value = validate_manifest(manifest)
    binding = {key: value[key] for key in SCOPE_FIELDS}
    binding.update(artifact_id=value["artifact_id"], transfer_id=value["transfer_id"],
                   execution_generation=str(value["execution_generation"]))
    binding_hash(binding)
    return binding


def validate_offer(payload: dict, route: str, desktop: str, origin: str) -> dict:
    if (not isinstance(payload, dict) or payload.get("type") != OFFER_TYPE
            or type(payload.get("version")) is not int or payload["version"] != 1):
        raise BlobError("invalid_artifact_blob_offer")
    if len(canonical(payload)) > MAX_CONTROL_BYTES:
        raise BlobError("artifact_blob_offer_too_large", 413)
    revision = payload.get("transport_revision")
    if type(revision) is not int or not 1 <= revision <= 2**53 - 1:
        raise BlobError("invalid_artifact_blob_transport_revision")
    value = validate_manifest(payload.get("manifest"))
    if value["client_route_id"] != route or value["desktop_id"] != desktop:
        raise BlobError("artifact_blob_route_mismatch", 409)
    offer = payload.get("blob_offer")
    if (not isinstance(offer, dict) or set(offer) != {"version", "relay", "private", "read_token"}
            or type(offer["version"]) is not int or offer["version"] != 1):
        raise BlobError("invalid_blob_offer")
    if offer["relay"] != relay_origin(origin):
        raise BlobError("artifact_blob_relay_mismatch", 409)
    checked_hex(offer["read_token"])
    private = validate_private_descriptor(offer["private"])
    if (private["binding_sha256"] != binding_hash(artifact_binding(value))
            or private["size"] != value["size_bytes"] or private["sha256"] != value["sha256"]):
        raise BlobError("artifact_blob_binding_mismatch", 409)
    return {"type": OFFER_TYPE, "version": 1, "manifest": value, "transport_revision": revision,
            "blob_offer": {**offer, "private": private}}


def stored_receipt(manifest: dict) -> dict:
    value = validate_manifest(manifest)
    return {"type": RECEIPT_TYPE, "version": 1, "status": "stored",
            **{key: value[key] for key in ("transfer_id", "client_route_id", "artifact_id", "sha256", "size_bytes")}}


def receipt_matches(manifest: dict, receipt: dict) -> bool:
    if not isinstance(receipt, dict):
        return False
    expected = stored_receipt(manifest)
    return (type(receipt.get("version")) is int and type(receipt.get("size_bytes")) is int
            and all(receipt.get(key) == value for key, value in expected.items()))
